### PR TITLE
Allow scatter and gather to broadcast

### PR DESCRIPTION
**Pull request by zasdfgbnm**
# Changes:

## 4. `gather`, `scatter`, `scatter_`, `scatter_add`, `scatter_add_` now support broadcasting.

The semantics is explained in `docs/source/notes/broadcasting.rst`. The docs of these operators are updated to reflect the change and point to `docs/source/notes/broadcasting.rst`.

Note:
`self.scatter(dim, index, src)` requires `self` and `src` to have the same number of dimensions. Further, both `self.scatter(dim, index, src)` and `self.gather(dim, index)` requires the number of dimensions of `index` <= number of dimensions of  `self`. When the number of dims in `index` is smaller than `self`, then `index` will be aligned to the left (instead of to the right as in broadcasting of other operators). My personal experience tells me that align to the left is more useful than to the right, and this behavior makes `torch.gather` get a very similar behavior to [`tensorflow.batch_gather`](https://www.tensorflow.org/versions/r1.13/api_docs/python/tf/batch_gather).

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
Hi @ezyang , this PR is now ready for review. It was a bit more tricky than I originally expected because I believe there is some bug in the current implementation. It's all explained in the PR message. The CI seems to be broken.

**Comment by ezyang** [comment]
Reading through the changes list, 1-3 all look good and desirable. Change 4 with left alignment makes me nervous but you may very well be right. @gchanan, I wonder if you have an idea here.

File: docs/source/notes/broadcasting.rst [hunk 1]
```diff
@@ -86,6 +86,42 @@ For Example::
     >>> (x.add_(y)).size()
     RuntimeError: The expanded size of the tensor (1) must match the existing size (7) at non-singleton dimension 2.
 
+Scatter and gather
+------------------
+
+``gather``, ``scatter_``, ``scatter_add_``, ``scatter``, and ``scatter_add`` also support broadcasting. But the broadcasting
+of these operators are more complicated than most other operators.
+
+When we do ``input.gather(dim, input)``, the following rules are applied:
```
Comment (ezyang) [inline docs/source/notes/broadcasting.rst]: Typo here, you mean index

File: docs/source/notes/broadcasting.rst [hunk 2]
```diff
@@ -86,6 +86,42 @@ For Example::
...
+
+- The number of dimensions of ``index`` must be equal to or smaller than the number of dimensions of ``input``.
+- If the number of dimensions of ``index`` is smaller than of ``input``, then ``index`` is aligned to the left,
```
Comment (ezyang) [inline docs/source/notes/broadcasting.rst]: If we do decide that left alignment is the way to go, it is worth calling out in super bold letters that this is what's happening here, and this is the primary difference between the broadcasting rules in the initial case and post case.

File: docs/source/notes/broadcasting.rst [hunk 3]
```diff
@@ -86,6 +86,42 @@ For Example::
...
+  that is, extra size 1 dimensions would be appended to the back to make them the same number of dimensions.
+- ``dim`` could be negative, if this is the case, then the ``dim`` is wrapped according to the original dimensions
+  if ``index``, not the broadcasted dimension. For example, if ``input`` is 8-dimensional and ``index`` is 4-dimensional,
+  then ``dim=-1`` is the same as ``dim=3``.
+- All dimensions other than the one specified by ``dim`` must either equal or one of them must be 1. If they are not equal,
+  then that dimesion would be expanded to the same size.
```
Comment (ezyang) [inline docs/source/notes/broadcasting.rst]: dimension

File: docs/source/notes/broadcasting.rst (hunk 2 above)
Comment (ezyang) [reply docs/source/notes/broadcasting.rst]: Since there is a lot of text here, it's easy to miss

File: docs/source/notes/broadcasting.rst [hunk 4]
```diff
@@ -86,6 +86,42 @@ For Example::
...
+``gather``, ``scatter_``, ``scatter_add_``, ``scatter``, and ``scatter_add`` also support broadcasting. But the broadcasting
+of these operators are more complicated than most other operators.
+
+When we do ``input.gather(dim, input)``, the following rules are applied:
+
+- The number of dimensions of ``index`` must be equal to or smaller than the number of dimensions of ``input``.
+- If the number of dimensions of ``index`` is smaller than of ``input``, then ``index`` is aligned to the left,
+  that is, extra size 1 dimensions would be appended to the back to make them the same number of dimensions.
+- ``dim`` could be negative, if this is the case, then the ``dim`` is wrapped according to the original dimensions
+  if ``index``, not the broadcasted dimension. For example, if ``input`` is 8-dimensional and ``index`` is 4-dimensional,
```
Comment (ezyang) [inline docs/source/notes/broadcasting.rst]: I think you meant "of `index`"

File: docs/source/notes/broadcasting.rst [hunk 5]
```diff
@@ -86,6 +86,42 @@ For Example::
...
+  then ``dim=-1`` is the same as ``dim=3``.
```
Comment (ezyang) [inline docs/source/notes/broadcasting.rst]: I find this mildly shocking.

Let's consider the `torch.gather` example.

```
out[i][j][k] = input[index[i][j][k]][j][k]  # if dim == 0
out[i][j][k] = input[i][index[i][j][k]][k]  # if dim == 1
out[i][j][k] = input[i][j][index[i][j][k]]  # if dim == 2
```

When index is broadcasted from 2D to 3D, we end up with equations like this:

```
out[i][j][k] = input[index[i][j]][j][k]  # if dim == 0
out[i][j][k] = input[i][index[i][j]][k]  # if dim == 1
out[i][j][k] = input[i][j][index[i][j]]  # if dim == 2
```

Now, let us consider the negative dimension cases. With your rule, I believe we get:

```
out[i][j][k] = input[index[i][j]][j][k]  # if dim == -2
out[i][j][k] = input[i][index[i][j]][k]  # if dim == -1
```

I find this incredibly shocking. The semantics of gather is "gather elements along the axis specified by dim." If I say -1, I expect this to refer to the axis of the INPUT I am gathering from, not the indexing tensor!

**Comment by ezyang** [comment]
@zasdfgbnm I am short on time today and cannot do a full review. However, if there is a way to split this patch, so that (1-3) are done before (4), that would be EXTREMELY helpful. I can probably get the first half merged much more quickly than part 4.

**Comment by LeanderK** [comment]
i think this PR will probably still take some time judging from the PRs it depends on?

**Comment by zasdfgbnm** [comment]
Yes

**Pull request closed without merging**